api/v1/mongodbcommunity_types.go

```diff
@@ -402,12 +423,39 @@ func (m MongoDBCommunity) GetScramOptions() scram.Options {
 		ignoreUnknownUsers = *m.Spec.Security.Authentication.IgnoreUnknownUsers
 	}
 
+	//Manage the Auth
+	listModes := m.Spec.Security.Authentication.Modes
+	containsDefaultMode := false
+	AutoAuthMechanism := ""
+	defaultSystemName, _ := ConvertAuthModeLabelToAuthModeSystemName(DefaultMode)
+
+	modesWithSystemNames := make([]string, len(listModes))
+
+	if len(listModes) == 0 {
+		AutoAuthMechanism = defaultSystemName
+		modesWithSystemNames = []string{AutoAuthMechanism}
+	} else {
+		for i, node := range listModes {
+			if v, ok := ConvertAuthModeLabelToAuthModeSystemName(node); ok {
+				modesWithSystemNames[i] = v
+				if v == defaultSystemName {
+					containsDefaultMode = true
+				}
+			}
+		}
+		if containsDefaultMode {
+			AutoAuthMechanism = defaultSystemName
+		} else {
+			AutoAuthMechanism, _ = ConvertAuthModeLabelToAuthModeSystemName(listModes[0])
+		}
+	}
+
```

Comment:
I have some concerts with this block. The variable names seem quite confusing, it's not immediately clear what everything is. Remember, we have AuthenticationModes (specified in CR) but also AuthenticationMechanisms (specified in automation config). It might be a good idea to distinguish between these two domains.

We appear to be a bit inconsistent with terminology used in the variable names (system/authMode/authMechanism) we should make sure that it is very clear wha is going on.

For example I was a bit confused by what `defaultSystemName`.  

So far in existing code we use `authentication.modes` to refer to the CR, and `authenticationMechanism` when dealing with the agent/automation config.

Reply:
I would also say that this block of code is not that easy to follow as currently written. I think ideally we want to keep our go code as flat as possible. In this case that would involve extracting out some well named helper functions to aid in readability. 

I left another comment below about actually needing a single entry in the modes array, with this being the case, we might be able to simplify this logic a bit! (please let me know if my understanding is correct and that we need at least one entry)

Reply:
Thanks for the reminder about the difference in the names with the CR and automation config. I simplified a bit the code. Let me know if you think it is still difficult to read.